edge-modules/load-gen/Program.cs

```diff
@@ -30,42 +32,45 @@ static async Task Main()
                     ModuleUtil.DefaultTransientRetryStrategy,
                     Logger);
 
-                using (var timers = new Timers())
-                {
-                    Guid batchId = Guid.NewGuid();
-                    Logger.LogInformation($"Batch Id={batchId}");
-
-                    // setup the message timer
-                    timers.Add(
-                        Settings.Current.MessageFrequency,
-                        Settings.Current.JitterFactor,
-                        () => GenerateMessageAsync(moduleClient, batchId));
-
-                    // setup the twin update timer
-                    timers.Add(
-                        Settings.Current.TwinUpdateFrequency,
-                        Settings.Current.JitterFactor,
-                        () => GenerateTwinUpdateAsync(moduleClient, batchId));
-
-                    timers.Start();
-                    (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
-                    Logger.LogInformation("Load gen running.");
-
-                    await cts.Token.WhenCanceled();
-                    Logger.LogInformation("Stopping timers.");
-                    timers.Stop();
-                    Logger.LogInformation("Closing connection to Edge Hub.");
-                    await moduleClient.CloseAsync();
-
-                    completed.Set();
-                    handler.ForEach(h => GC.KeepAlive(h));
-                    Logger.LogInformation("Load Gen complete. Exiting.");
-                }
+                Guid batchId = Guid.NewGuid();
+                Logger.LogInformation($"Batch Id={batchId}");
+
+                // setup the message PeriodicTask
+                messageTask = Option.Some(new PeriodicTask(
```

Comment:
I don't understand how periodic task works w.r.t its cancellation token. I do understand that the ShutdownHandler creates a cancellation token and hooks cancellation into typical "shutdown" actions. It seems that PeriodicTask has it's cancellation token configured to cancel when the object is disposed. My guess is that this works because the underlying task runs async, allowing for the main() to exit then trigger disposal?

Reply:
wrap the tasks around the Shutdown handler cancellation token await.

Reply:
Can you explain more what you mean @varunpuranik ?

Reply:
Put the PeriodicTask in a using statement with an await for the cancellation token cancel